### PR TITLE
[IA-4465] Cleanup PD decoration

### PR DESCRIPTION
**Pull request by LizBaldo**
### What
- Make use of the decorated PD throughout the UI codebase

### Why
- Having the API return a consistent data structure instead of transforming to and from it multiple time should reduce the risk of bugs significantly

### Testing strategy
- [ ]

### CODE REVIEW

File: src/libs/ajax/leonardo/Disks.ts [hunk 1]
```diff
@@ -13,12 +13,13 @@ export const Disks = (signal) => {
   });
 
   const v1Func = () => ({
-    list: async (labels = {}): Promise<ListDiskItem[]> => {
+    list: async (labels = {}): Promise<DecoratedPersistentDisk[]> => {
```
Comment (mlilynolting) [inline src/libs/ajax/leonardo/Disks.ts]: This is where Nick and Miguel would be proposing a change.

I don't know if we want to implement this in a way that's fully compliant with the recommendation of the cross-team proposal, but for the future, here's my understanding of what that would entail:

In [the long discussion on the other sanitization PR](https://github.com/DataBiosphere/terra-ui/pull/4110), to fit with the structure proposed in the [code sharing proposal](https://docs.google.com/document/d/1vefkF3tbIPgn_2Kbx6lscFK6nYoZoaowlkXXkszxijA/edit#heading=h.cmco44cpjoai), this method would remain unchanged, returning `ListDiskItem[]`. This Ajax code is the Data Client layer.

Above that we would add two new classes: a Data Provider and a State Provider. Above that would be the Components, which would interact only with State Providers. (Right now our Components are directly calling Ajax.)

This is how I understand these new layers to work. The best example I can find is `AnalysisFile`, which is a type used in the Component `ExportAnalysisModal.ts`. It's unfortunately not a perfectly straightforward example, and I'm sure if I were better at React I'd be able to explain it more simply, but for now this is the best I can do.

So in the case of the AnalysisFile[] data type:

### Data Client fetches from the server with Ajax calls
* Both `GoogleStorage.listAnalyses()` and `AzureStorage.listNotebooks()` return AnalysisFile[] objects. This is the raw data from the server. The Data Client layer makes raw HTTP calls and handles headers, JSON, and responses, returning unsanitized objects.

### Provider layer wraps the Data Client
* `AnalysisProvider` is responsible for calling the Data Client (the Ajax methods). It has a `list` method which takes params and returns `AnalysisFile[]`. This is where the data is sanitized - the Ajax list actions (listAnalyses and listNotebooks) return GCP and Azure files, and the Provider unifies them as AnalysisFiles. The State Provider only needs to know about AnalysisFiles, and doesn't need any logic to distinguish GCP from Azure.

### State Provider layer wraps the Provider
* `useAnalysisExportState` sets up some of the wiring needed around the Ajax fetch, like `signal`, metrics, and tracking whether requests are in progress (using the LoadedState wrapper). Most importantly it creates stateful data holders (`existingAnalysisFiles`), and updates them whenever it asks its Provider (`AnalysisProvider`) for data.
* `useState` is the [React hook](https://react.dev/learn/state-a-components-memory) that enables this - providing a getter and setter for values that we want to preserve across components.

### Component layer wraps the State Provider
* When it initializes, `ExportAnalysisModal` asks `useAnalysisExportState`, its State Provider, for the `existingAnalysisFiles` state. This starts off empty. When the modal is ready to get files for a workspace (as part of the copy workflow which isn't important for this example), it calls a method on the State Provider which causes `existingAnalysisFiles` to update. The modal (Component) is watching that state, and rerenders on change.

So, if we were to follow this model for disks, we'd want components like WorkspaceContainer to stop calling `Disks.disksV1().list()` directly. Instead, the component would ask a state provider `usePersistentDisk` for a stateful collection of disks. The state provider would ask a data provider `DiskProvider` for pretty disks. And the data provider would call `Disks.v1func().list()`.

Maybe there's some kind of intermediate structure we could set up with a TODO to expand into the full architecture. We didn't scope out a refactor on this scale.

**Comment by mlilynolting** [comment]
It's amazing to see so many helper functions deleted and the imports simplified so much! This PR is a joy

**Comment by mlilynolting** [comment]
Would it be possible to change `DecoratedPersistentDisk` to `PersistentDisk` and `PersistentDisk` to, maybe, `RawPersistentDisk`? If everything across the client is using one PD type then it probably isn't important to call it Decorated.

File: src/analysis/Environments.test.ts [hunk 2]
```diff
@@ -621,7 +621,7 @@ describe('Environments', () => {
   describe('Disks - ', () => {
     it('Renders page correctly with a disk', async () => {
       const disk = generateTestDiskWithGoogleWorkspace();
-      asMockedFn(list).mockReturnValue(Promise.resolve([disk]));
+      asMockedFn(list).mockReturnValue(Promise.resolve(disk));
```
Comment (jdcanas) [inline src/analysis/Environments.test.ts]: Hmm.. this doesn't seem correct, list disks should still return a list?

File: src/libs/ajax/leonardo/Disks.ts (hunk 1 above)
Comment (jdcanas) [reply src/libs/ajax/leonardo/Disks.ts]: Lets chat about this in walkthrough?

**Review by jdcanas — APPROVED** [review]

File: src/analysis/Environments.test.ts (hunk 2 above)
Comment (LizBaldo) [reply src/analysis/Environments.test.ts]: Oh yeah, my bad! I was trying a bunch of stuff to make the test pass and did not clean up 🙈

**Comment by LizBaldo** [comment]
@jdcanas and @mlilynolting, even though all unit and integration tests are passing, my testing shows that PDs are not being displayed by the UI correctly in the compute and cloud environment modals. 
Maybe we could pair again? I would like to write the necessary tests to cover that gap, and then fix the code, this is a bit scary 😅

**Comment by LizBaldo** [comment]
@jdcanas and @mlilynolting this is ready for your re-review :) I added a unit test to catch the bug that I saw when testing things manually on a BEE and things are looking good now.

File: src/libs/ajax/leonardo/Disks.ts [hunk 3]
```diff
@@ -13,12 +13,13 @@ export const Disks = (signal) => {
...
   const v1Func = () => ({
-    list: async (labels = {}): Promise<ListDiskItem[]> => {
+    list: async (labels = {}): Promise<PersistentDisk[]> => {
       const res = await fetchLeo(
         `api/google/v1/disks${qs.stringify(labels, { addQueryPrefix: true })}`,
         _.mergeAll([authOpts(), appIdentifier, { signal }])
       );
-      return res.json();
+      const disks: ListDiskItem[] = await res.json();
+      return mapToPdTypes(disks);
```
Comment (lucymcnatt) [inline src/libs/ajax/leonardo/Disks.ts:22]: so this `mapToPdTypes` function takes in `disks: RawPersistentDisk[]` from what I see in disk-utils, but this is saying that disks is a `ListDiskItem[]` are those two types equivalent?

Comment (lucymcnatt) [reply src/libs/ajax/leonardo/Disks.ts:22]: wait I just scrolled further and saw that they are, nvmd

File: src/libs/ajax/leonardo/models/disk-models.ts [hunk 4]
```diff
@@ -108,9 +107,10 @@ export const googlePdTypes: Record<PDLabels, GooglePdType> = {
     regionToPricesName: 'monthlySSDDiskPrice',
   },
 };
-export type DecoratedPersistentDisk = {
+export type PersistentDisk = {
   diskType: GooglePdType;
-} & Omit<PersistentDisk, 'diskType'>;
+} & Omit<RawPersistentDisk, 'diskType'>;
```
Comment (lucymcnatt) [inline src/libs/ajax/leonardo/models/disk-models.ts:112]: just curious, what does the `Omit<>` do here?

**Review by lucymcnatt — APPROVED** [review]

File: src/libs/ajax/leonardo/models/disk-models.ts (hunk 4 above)
Comment (LizBaldo) [reply src/libs/ajax/leonardo/models/disk-models.ts:112]: My understanding is that it removes the original diskType from the object so it can be replaced by the GooglePdType

**Review by cahrens — APPROVED** [review]
reviewed workspace file only

File: src/analysis/utils/disk-utils.ts [hunk 5]
```diff
@@ -37,18 +37,11 @@ export const pdTypeFromDiskType = (type: GoogleDiskType): GooglePdType =>
      */
   ) as GooglePdType; // TODO: Remove cast
 
-export const updatePdType = (disk: PersistentDisk): DecoratedPersistentDisk => ({
+export const updatePdType = (disk: RawPersistentDisk): PersistentDisk => ({
```
Comment (mlilynolting) [inline src/analysis/utils/disk-utils.ts:40]: nit: for naming, this could be treated as a black box that takes a RawPD and returns a PD - it could be called `toPersistentDisk` or `asPersistentDisk`

File: src/analysis/utils/disk-utils.ts [hunk 6]
```diff
@@ -37,18 +37,11 @@ export const pdTypeFromDiskType = (type: GoogleDiskType): GooglePdType =>
...
 });
-export const mapToPdTypes = (disks: PersistentDisk[]): DecoratedPersistentDisk[] => _.map(updatePdType, disks);
-
-export const undecoratePd = (disk: DecoratedPersistentDisk): PersistentDisk => ({
-  ...disk,
-  diskType: disk.diskType.value,
-});
-
-export const mapToUndecoratedPds = (disks: DecoratedPersistentDisk[]): PersistentDisk[] => _.map(undecoratePd, disks);
+export const mapToPdTypes = (disks: RawPersistentDisk[]): PersistentDisk[] => _.map(updatePdType, disks);
```
Comment (mlilynolting) [inline src/analysis/utils/disk-utils.ts:44]: nit: this could follow the above and be `toPersistentDisks` or `asPersistentDisks`

File: src/analysis/utils/cost-utils.ts [hunk 7]
```diff
@@ -282,9 +280,7 @@ export const getPersistentDiskCostMonthly = (disk: DecoratedPersistentDisk, comp
 };
 
 export const getPersistentDiskCostHourly = (
-  disk:
-    | DecoratedPersistentDisk
-    | { size: number; status: LeoDiskStatus; diskType: GooglePdType; cloudContext?: CloudContext },
+  disk: PersistentDisk | { size: number; status: LeoDiskStatus; diskType: GooglePdType; cloudContext?: CloudContext },
```
Comment (mlilynolting) [inline src/analysis/utils/cost-utils.ts:283]: nit: if you only need these 4 fields off the PD type, you could do `disk: Pick<PersistentDisk, ['size', 'status', 'diskType', 'cloudContext']>`

File: src/analysis/utils/cost-utils.ts [hunk 8]
```diff
@@ -213,7 +211,7 @@ export const getGalaxyComputeCost = (app: App): number => {
  */
 export const getGalaxyDiskCost = (
   disk:
-    | DecoratedPersistentDisk
+    | PersistentDisk
```
Comment (mlilynolting) [inline src/analysis/utils/cost-utils.ts:214]: nit: this could also be a `Pick<`, an even shorter one

Comment (mlilynolting) [reply src/analysis/utils/cost-utils.ts:214]: Also, I realized I don't know what `key: | Type` means - does the initial ` | ` do something? Or is it just sugar to make the types look good on multiple lines?
